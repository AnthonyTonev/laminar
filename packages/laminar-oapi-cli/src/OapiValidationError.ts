export class OapiValidationError extends Error {
  public readonly errors: string[];

  public constructor(message: string, errors: string[] = []) {
    super(message);
    this.errors = errors;
  }

  public toString(): string {
    return `Error: ${this.message}\n${this.errors.map((item) => ` | ${item}`).join('\n')}\n`;
  }
}
