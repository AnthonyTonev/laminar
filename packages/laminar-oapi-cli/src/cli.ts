#!/usr/bin/env node

import { laminarOapi } from './laminar-oapi';

laminarOapi().parse(process.argv);
