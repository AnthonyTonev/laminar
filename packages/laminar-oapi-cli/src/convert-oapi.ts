import {
  document,
  Document,
  mapWithContext,
  Type,
  withIdentifier,
  withImports,
} from '@ovotech/ts-compose';
import {
  OpenAPIObject,
  ParameterObject,
  PathItemObject,
  ReferenceObject,
  RequestBodyObject,
  ResponseObject,
  ResponsesObject,
  SchemaObject,
} from 'openapi3-ts';
import * as ts from 'typescript';
import { convertSchema } from './convert-schema';
import {
  AstContext,
  isSchemaObject,
  isResponseObject,
  getReferencedObject,
  isRequestBodyObject,
  isParameterObject,
} from './traverse';

interface AstParameters {
  in: {
    query?: ts.TypeElement[];
    headers?: ts.TypeElement[];
    path?: ts.TypeElement[];
    cookies?: ts.TypeElement[];
  };
  context: AstContext;
}

const title = (str: string): string => str.replace(/^./, first => first.toUpperCase());
const documentation = (summary?: string, description?: string): string | undefined =>
  summary || description
    ? [summary, description].filter(item => item !== undefined).join('\n')
    : undefined;
const cleanIdentifierName = (str: string): string => str.replace(/[^0-9a-zA-Z_$]+/g, '');

const pathToIdentifier = (path: string): string =>
  'T' +
  path
    .split('/')
    .map(cleanIdentifierName)
    .map(title)
    .join('');

const toParamLocation = (
  location: 'header' | 'cookie' | 'path' | 'query',
): 'headers' | 'cookies' | 'path' | 'query' => {
  switch (location) {
    case 'header':
      return 'headers';
    case 'cookie':
      return 'cookies';
    default:
      return location;
  }
};

const toParamName = (location: 'header' | 'cookie' | 'path' | 'query', name: string): string =>
  location === 'header' ? name.toLowerCase() : name;

const convertParameters = (
  context: AstContext,
  parameters: (ParameterObject | ReferenceObject)[],
): Document<ts.TypeLiteralNode, AstContext> => {
  const astParams = parameters.reduce<AstParameters>(
    (node, paramOrRef) => {
      const param = getReferencedObject(paramOrRef, isParameterObject, node.context);
      const paramNode = param.schema
        ? convertSchema(node.context, param.schema)
        : document(node.context, Type.Any);
      const params = node.in[toParamLocation(param.in)] || [];
      return {
        context: paramNode.context,
        in: {
          ...node.in,
          [toParamLocation(param.in)]: [
            ...params,
            Type.Prop({
              name: toParamName(param.in, param.name),
              jsDoc: documentation(param.description),
              type: paramNode.type,
              isOptional: !param.required,
            }),
          ],
        },
      };
    },
    { in: {}, context },
  );

  return {
    type: Type.TypeLiteral({
      props: Object.entries(astParams.in).map(([name, items]) =>
        Type.Prop({ name, type: Type.TypeLiteral({ props: items }) }),
      ),
    }),

    context: astParams.context,
  };
};

/**
 * If root repsponse is string, we allow it to be readable stream as well
 */
const convertResponse = (
  context: AstContext,
  schemaOrRef?: SchemaObject | ReferenceObject,
): Document<ts.TypeNode | undefined, AstContext> => {
  if (schemaOrRef) {
    const schema = getReferencedObject(schemaOrRef, isSchemaObject, context);
    const responseDocument = convertSchema(context, schemaOrRef);
    if (schema.type === 'string') {
      return document(
        withImports(responseDocument.context, { module: 'stream', named: [{ name: 'Readable' }] }),
        Type.Union([responseDocument.type, Type.Referance('Readable')]),
      );
    } else {
      return responseDocument;
    }
  } else {
    return document(context, undefined);
  }
};

const convertResponses = (
  context: AstContext,
  name: string,
  responses: ResponsesObject,
): Document<ts.TypeReferenceNode, AstContext> => {
  const responseEntries = Object.values<ResponseObject | ReferenceObject>(responses);

  const params = mapWithContext(context, responseEntries, (responseContext, responseOrRef) => {
    const response = getReferencedObject(responseOrRef, isResponseObject, responseContext);
    const schema =
      response?.content?.['application/json']?.schema ?? response?.content?.['*/*']?.schema;

    const node = convertResponse(responseContext, schema);

    const nodeContext = withImports(node.context, {
      module: '@ovotech/laminar',
      named: [{ name: 'LaminarResponse' }],
    });

    if (node.type !== undefined) {
      return document(nodeContext, [Type.Referance('LaminarResponse', [node.type]), node.type]);
    } else {
      return document(nodeContext, [Type.Referance('LaminarResponse')]);
    }
  });

  const responseTypes = params.items.reduce((acc, item) => [...acc, ...item]);

  return responseTypes.length
    ? document(
        withIdentifier(
          params.context,
          Type.Alias({
            name,
            type: Type.Union([
              ...responseTypes,
              Type.Referance('Promise', [Type.Union(responseTypes)]),
            ]),
            isExport: true,
          }),
        ),
        Type.Referance(name),
      )
    : document(
        withImports(context, { module: '@ovotech/laminar', named: [{ name: 'ResolverResponse' }] }),
        Type.Referance('ResolverResponse'),
      );
};

const convertRequestBody = (
  context: AstContext,
  requestBodyOrRef: RequestBodyObject | ReferenceObject,
): Document<ts.TypeLiteralNode, AstContext> => {
  const requestBody = getReferencedObject(requestBodyOrRef, isRequestBodyObject, context);
  const schema = requestBody.content['application/json']
    ? requestBody.content['application/json'].schema
    : {};
  const node = schema ? convertSchema(context, schema) : document(context, Type.Any);
  return document(
    node.context,
    Type.TypeLiteral({
      props: [Type.Prop({ name: 'body', type: node.type, isOptional: !requestBody.required })],
    }),
  );
};

export const convertOapi = (
  context: AstContext,
  api: OpenAPIObject,
): Document<ts.InterfaceDeclaration, AstContext> => {
  const paths = mapWithContext(
    context,
    Object.entries<PathItemObject>(api.paths),
    (pathContext, [path, pathApiOrRef]) => {
      const { parameters, description, summary, ...methodsApiOrRef } = pathApiOrRef;
      const pathApi = getReferencedObject(methodsApiOrRef, isSchemaObject, context);

      const methods = mapWithContext(
        pathContext,
        Object.entries(pathApi),
        (methodContext, [method, operation]) => {
          const astParams =
            operation.parameters || parameters
              ? convertParameters(methodContext, [
                  ...(parameters ?? []),
                  ...(operation.parameters ?? []),
                ])
              : document(methodContext, Type.TypeLiteral());

          const astRequestBody = operation.requestBody
            ? convertRequestBody(astParams.context, operation.requestBody)
            : document(methodContext, Type.TypeLiteral());

          const identifier = pathToIdentifier(path) + title(method);
          const contextIdentifier = identifier + 'Context';
          const responseIdentifier = identifier + 'Response';
          const doc = documentation(
            operation.summary || summary,
            operation.description || description,
          );
          const contextInterface = Type.Interface({
            name: contextIdentifier,
            isExport: true,
            jsDoc: doc,
            ext: [{ name: 'Context' }, { name: 'OapiContext' }],
            props: astParams.type.members.concat(astRequestBody.type.members),
          });

          const responseAst = convertResponses(
            astRequestBody.context,
            responseIdentifier,
            operation.responses,
          );

          const methodCall = Type.Arrow(
            [
              Type.Param({
                name: 'context',
                type: Type.Intersection([Type.Referance(contextIdentifier), Type.Referance('C')]),
              }),
            ],
            responseAst.type,
          );

          const methodSignature = Type.Prop({
            name: method,
            type: methodCall,
            jsDoc: operation.description,
          });

          return document(withIdentifier(responseAst.context, contextInterface), methodSignature);
        },
      );

      return document(
        methods.context,
        Type.Prop({
          name: Type.LiteralString(path),
          type: Type.TypeLiteral({ props: methods.items }),
        }),
      );
    },
  );

  return document(
    withImports(
      withImports(paths.context, {
        module: '@ovotech/laminar-oapi',
        named: [
          { name: 'OapiContext' },
          { name: 'OapiConfig' },
          ...(api.components && api.components.securitySchemes
            ? [{ name: 'OapiSecurityResolver' }]
            : []),
        ],
      }),
      { module: '@ovotech/laminar', named: [{ name: 'Context' }] },
    ),
    Type.Interface({
      name: 'Config',
      isExport: true,
      ext: [{ name: 'OapiConfig', types: [Type.Referance('C')] }],
      typeArgs: [
        Type.TypeArg({
          name: 'C',
          ext: Type.TypeLiteral(),
          defaultType: Type.TypeLiteral(),
        }),
      ],
      props: [
        Type.Prop({ name: 'paths', type: Type.TypeLiteral({ props: paths.items }) }),
        ...(api.components && api.components.securitySchemes
          ? [
              Type.Prop({
                name: 'security',
                type: Type.TypeLiteral({
                  props: Object.keys(api.components.securitySchemes).map(scheme =>
                    Type.Prop({
                      name: scheme,
                      type: Type.Referance('OapiSecurityResolver', [Type.Referance('C')]),
                    }),
                  ),
                }),
              }),
            ]
          : []),
      ],
    }),
  );
};
