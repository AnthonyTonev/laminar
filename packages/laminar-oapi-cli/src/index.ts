export { convertSchema } from './convert-schema';
export { convertOapi } from './convert-oapi';
export { oapiTs, schemaTs } from './convert';
