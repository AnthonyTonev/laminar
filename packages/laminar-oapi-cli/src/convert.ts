import { compile, Schema } from '@ovotech/json-schema';
import { printDocument } from '@ovotech/ts-compose';
import { OpenAPIObject, SchemaObject } from 'openapi3-ts';
import { convertOapi } from './convert-oapi';
import { convertSchema } from './convert-schema';
import { OapiValidationError } from './OapiValidationError';

export const isOpenApiObject = (schema: Schema): schema is OpenAPIObject =>
  typeof schema === 'object' &&
  schema !== null &&
  'openapi' in schema &&
  'info' in schema &&
  'paths' in schema;

export const oapiTs = async (api: OpenAPIObject | string): Promise<string> => {
  const { schema, refs } = await compile(api);
  if (!isOpenApiObject(schema)) {
    throw new OapiValidationError('Invalid Schema', [
      'Schema could not be converted to typescript. Not a json-schema',
    ]);
  }

  const context = { root: schema, refs };

  return printDocument(convertOapi(context, schema));
};

export const schemaTs = async (api: Schema | string): Promise<string> => {
  const { schema, refs } = await compile(api);

  return printDocument(convertSchema({ root: schema as SchemaObject, refs }, schema));
};
