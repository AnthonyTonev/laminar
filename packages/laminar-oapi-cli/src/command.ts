import { compile, validate } from '@ovotech/json-schema';
import { Schema } from '@ovotech/json-refs';
import { Command } from 'commander';
import * as fs from 'fs';
import { openapiV3 } from 'openapi-schemas';
import { oapiTs } from './convert';
import { OapiValidationError } from './OapiValidationError';

export const processFile = async (
  fileName: string,
): Promise<{ content: string; uris: string[] }> => {
  const { schema, uris } = await compile(fileName);
  const check = await validate(openapiV3 as Schema, schema);

  if (!check.valid) {
    throw new OapiValidationError('Invalid API Definition', check.errors);
  }

  return { content: await oapiTs(fileName), uris };
};

export const toFiles = (uris: string[]): string[] =>
  uris.filter(uri => uri.startsWith('file://')).map(uri => uri.substring('file://'.length));

new Command()
  .description('Convert openapi schemas to typescript types')
  .option('-w, --watch', 'Watch for file changes and update live')
  .arguments('<input-file> <output-file>')
  .action(async (inputFile, outputFile, { watch }) => {
    const { content, uris } = await processFile(inputFile);
    fs.writeFileSync(outputFile, content);
    process.stdout.write(`Conterted ${inputFile} -> ${outputFile}\n`);

    if (watch) {
      process.stdout.write(`Watching ${toFiles(uris).join(', ')} for changes\n`);

      toFiles(uris).forEach(file =>
        fs.watchFile(file, async () => {
          try {
            const update = await processFile(inputFile);
            fs.writeFileSync(outputFile, update.content);
            process.stdout.write(`Updated ${inputFile} -> ${outputFile} (trigger ${file})\n`);
          } catch (error) {
            if (error instanceof OapiValidationError) {
              process.stderr.write(`-----\n${error.toString()}`);
            } else {
              throw error;
            }
          }
        }),
      );
    }
  })
  .parse(process.argv);
