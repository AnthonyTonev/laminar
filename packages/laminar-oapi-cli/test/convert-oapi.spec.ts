import { readdirSync, readFileSync } from 'fs';
import { safeLoad } from 'js-yaml';
import { OpenAPIObject } from 'openapi3-ts';
import { join } from 'path';
import { oapiTs } from '../src';

const oapiSchemas = readdirSync(join(__dirname, 'specs'))
  .filter(file => file.endsWith('.yaml'))
  .map<[string, OpenAPIObject]>(file => [
    file,
    safeLoad(String(readFileSync(join(__dirname, 'specs', file)))),
  ]);

describe('Json Schema Ts', () => {
  it.each(oapiSchemas)('Test %s', async (file, schema) => {
    const result = await oapiTs(schema);
    expect(result).toMatchSnapshot();
  });
});
