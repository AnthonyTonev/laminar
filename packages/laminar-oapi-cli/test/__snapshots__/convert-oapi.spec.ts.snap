// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Json Schema Ts Test api-with-examples.yaml 1`] = `
"import { ResolverResponse, RouteContext } from \\"@ovotech/laminar\\";

import { OapiContext, OapiPaths, OapiSecurityResolver, OapiSecurityResolvers } from \\"@ovotech/laminar-oapi\\";

export interface Paths<TContext = {}> extends OapiPaths<TContext> {
    \\"/\\": {
        get: (context: GetContext & TContext) => ResolverResponse;
    };
    \\"/v2\\": {
        get: (context: V2GetContext & TContext) => ResolverResponse;
    };
}

export interface GetContext extends OapiContext, RouteContext {
}

export interface V2GetContext extends OapiContext, RouteContext {
}"
`;

exports[`Json Schema Ts Test balance-service.yaml 1`] = `
"import { LaminarResponse, ResolverResponse, RouteContext } from \\"@ovotech/laminar\\";

import { OapiContext, OapiPaths, OapiSecurityResolver, OapiSecurityResolvers } from \\"@ovotech/laminar-oapi\\";

export interface Paths<TContext = {}> extends OapiPaths<TContext> {
    \\"/internal/status\\": {
        /**
         * Returns 200 if the application is up. It does not do deep check. Always gives back 200.
         */
        get: (context: InternalStatusGetContext & TContext) => ResolverResponse;
    };
    \\"/internal/swagger.yaml\\": {
        get: (context: InternalSwaggeryamlGetContext & TContext) => ResolverResponse;
    };
    \\"/api/v2/accounts/{accountId}/balance\\": {
        /**
         * Returns the balance
         */
        get: (context: ApiV2AccountsAccountIdBalanceGetContext & TContext) => ApiV2AccountsAccountIdBalanceGetResponse;
    };
    \\"/api/v2/accounts/{accountId}/transactions\\": {
        /**
         * Returns the transaction history
         */
        get: (context: ApiV2AccountsAccountIdTransactionsGetContext & TContext) => ApiV2AccountsAccountIdTransactionsGetResponse;
    };
    \\"/api/v2/accounts/{accountId}/adjust-balance\\": {
        /**
         * Returns with the adjusted balance
         */
        post: (context: ApiV2AccountsAccountIdAdjustbalancePostContext & TContext) => ApiV2AccountsAccountIdAdjustbalancePostResponse;
    };
    \\"/api/v3/accounts/{accountId}/balance\\": {
        /**
         * Returns the balance
         */
        get: (context: ApiV3AccountsAccountIdBalanceGetContext & TContext) => ApiV3AccountsAccountIdBalanceGetResponse;
    };
    \\"/api/v3/accounts/{accountId}/transactions\\": {
        /**
         * Returns the transaction history
         */
        get: (context: ApiV3AccountsAccountIdTransactionsGetContext & TContext) => ApiV3AccountsAccountIdTransactionsGetResponse;
    };
    \\"/api/v3/accounts/{accountId}/adjust-balance\\": {
        /**
         * Returns with the adjusted balance
         */
        post: (context: ApiV3AccountsAccountIdAdjustbalancePostContext & TContext) => ApiV3AccountsAccountIdAdjustbalancePostResponse;
    };
}

export interface InternalStatusGetContext extends OapiContext, RouteContext {
}

export interface InternalSwaggeryamlGetContext extends OapiContext, RouteContext {
}

export interface BalanceSummaryHttpResponse {
    balance?: number;
    /**
     * Last transaction date-time which updated the balance
     */
    effectiveAt?: string;
    /**
     * When balance service updated the balance
     */
    updatedAt?: string;
    [key: string]: any;
}

export interface HttpErrorResponse {
    /**
     * PLANNED. Currently not sent
     */
    errorCode?: string;
    message?: string;
    [key: string]: any;
}

export type ApiV2AccountsAccountIdBalanceGetResponse = (BalanceSummaryHttpResponse | LaminarResponse<BalanceSummaryHttpResponse>) | LaminarResponse<HttpErrorResponse> | LaminarResponse<HttpErrorResponse>;

export interface ApiV2AccountsAccountIdBalanceGetContext extends OapiContext, RouteContext {
    path: {
        accountId: string;
    };
}

export interface TransactionHistoryHttpResponseV2 {
    summary?: TransactionHistorySummary;
    /**
     * List of transactions ordered by transactionDateTime in ascending order.
     */
    transactions?: (GenericTransactionEntity | ChargeTransactionEntity)[];
    [key: string]: any;
}

export interface TransactionHistorySummary {
    /**
     * Number of search hits
     */
    totalSize?: number;
    [key: string]: any;
}

export interface GenericTransactionEntity {
    transactionType?: \\"TopUp\\" | \\"BalanceAdjustment\\";
    amount?: number;
    transactionDateTime?: string;
    /**
     * CamelCased cause why the adjustment is needed, e.g. 'GoodWill', 'WarmHomeDiscount', 'DiscretionaryCredit', 'AppPayment'
     */
    cause?: string;
    [key: string]: any;
}

export interface ChargeTransactionEntity {
    transactionType?: \\"ElectricityUsageCharge\\" | \\"ElectricityStandingCharge\\";
    amount?: number;
    transactionDateTime?: string;
    fuelType?: \\"Electricity\\" | \\"Gas\\";
    [key: string]: any;
}

export type ApiV2AccountsAccountIdTransactionsGetResponse = (TransactionHistoryHttpResponseV2 | LaminarResponse<TransactionHistoryHttpResponseV2>) | LaminarResponse<HttpErrorResponse> | LaminarResponse<HttpErrorResponse>;

export interface ApiV2AccountsAccountIdTransactionsGetContext extends OapiContext, RouteContext {
    path: {
        accountId: string;
    };
    headers: {
        scenario?: string;
    };
    cookies: {
        session?: string;
    };
    query: {
        fromDate: string;
        toDate: string;
        transactionType?: (\\"TopUp\\" | \\"ElectricityUsageCharge\\" | \\"ElectricityStandingCharge\\" | \\"BalanceAdjustment\\")[];
    };
}

export interface AdjustBalanceRequest {
    transactionId?: string;
    /**
     * Signed amount to adjust the balance with
     */
    amount?: number;
    /**
     * Adjustment date-time
     */
    transactionDateTime?: string;
    /**
     * CamelCased cause why the adjustment is needed, e.g. 'GoodWill', 'WarmHomeDiscount', 'DiscretionaryCredit'
     */
    cause?: string;
    [key: string]: any;
}

export type ApiV2AccountsAccountIdAdjustbalancePostResponse = (BalanceSummaryHttpResponse | LaminarResponse<BalanceSummaryHttpResponse>) | LaminarResponse<HttpErrorResponse> | LaminarResponse<HttpErrorResponse>;

export interface ApiV2AccountsAccountIdAdjustbalancePostContext extends OapiContext, RouteContext {
    path: {
        accountId: string;
    };
    body: AdjustBalanceRequest;
}

export type ApiV3AccountsAccountIdBalanceGetResponse = (BalanceSummaryHttpResponse | LaminarResponse<BalanceSummaryHttpResponse>) | LaminarResponse<HttpErrorResponse> | LaminarResponse<HttpErrorResponse>;

export interface ApiV3AccountsAccountIdBalanceGetContext extends OapiContext, RouteContext {
    path: {
        accountId: string;
    };
}

export interface TransactionHistoryHttpResponseV3 {
    summary?: TransactionHistorySummary;
    /**
     * List of transactions ordered by transactionDateTime in ascending order.
     */
    transactions?: (GenericTransactionEntity | UsageChargeTransactionEntity | StandingChargeTransactionEntity)[];
    [key: string]: any;
}

export interface UsageChargeTransactionEntity {
    transactionType?: \\"ElectricityUsageCharge\\";
    amount?: number;
    transactionDateTime?: string;
    fuelType?: \\"Electricity\\" | \\"Gas\\";
    /**
     * Meter Serial Number
     */
    msn?: string;
    /**
     * The meter read that triggered the charge.
     */
    read?: number;
    /**
     * The consumption amount (kwH) corresponding to this charge.
     */
    consumption?: number;
    [key: string]: any;
}

export interface StandingChargeTransactionEntity {
    transactionType?: \\"ElectricityStandingCharge\\";
    amount?: number;
    transactionDateTime?: string;
    fuelType?: \\"Electricity\\" | \\"Gas\\";
    [key: string]: any;
}

export type ApiV3AccountsAccountIdTransactionsGetResponse = (TransactionHistoryHttpResponseV3 | LaminarResponse<TransactionHistoryHttpResponseV3>) | LaminarResponse<HttpErrorResponse> | LaminarResponse<HttpErrorResponse>;

export interface ApiV3AccountsAccountIdTransactionsGetContext extends OapiContext, RouteContext {
    path: {
        accountId: string;
    };
    query: {
        fromDate: string;
        toDate: string;
        transactionType?: (\\"TopUp\\" | \\"ElectricityUsageCharge\\" | \\"ElectricityStandingCharge\\" | \\"BalanceAdjustment\\")[];
    };
}

export type ApiV3AccountsAccountIdAdjustbalancePostResponse = (BalanceSummaryHttpResponse | LaminarResponse<BalanceSummaryHttpResponse>) | LaminarResponse<HttpErrorResponse> | LaminarResponse<HttpErrorResponse>;

export interface ApiV3AccountsAccountIdAdjustbalancePostContext extends OapiContext, RouteContext {
    path: {
        accountId: string;
    };
    body: AdjustBalanceRequest;
}

export interface SecurityResolvers<TContext = {}> extends OapiSecurityResolvers<TContext> {
    BasicAuth: OapiSecurityResolver;
    BearerAuth: OapiSecurityResolver;
    ApiKeyAuth: OapiSecurityResolver;
}"
`;

exports[`Json Schema Ts Test link-example.yaml 1`] = `
"import { LaminarResponse, ResolverResponse, RouteContext } from \\"@ovotech/laminar\\";

import { OapiContext, OapiPaths, OapiSecurityResolver, OapiSecurityResolvers } from \\"@ovotech/laminar-oapi\\";

export interface Paths<TContext = {}> extends OapiPaths<TContext> {
    \\"/2.0/users/{username}\\": {
        get: (context: 20UsersUsernameGetContext & TContext) => 20UsersUsernameGetResponse;
    };
    \\"/2.0/repositories/{username}\\": {
        get: (context: 20RepositoriesUsernameGetContext & TContext) => 20RepositoriesUsernameGetResponse;
    };
    \\"/2.0/repositories/{username}/{slug}\\": {
        get: (context: 20RepositoriesUsernameSlugGetContext & TContext) => 20RepositoriesUsernameSlugGetResponse;
    };
    \\"/2.0/repositories/{username}/{slug}/pullrequests\\": {
        get: (context: 20RepositoriesUsernameSlugPullrequestsGetContext & TContext) => 20RepositoriesUsernameSlugPullrequestsGetResponse;
    };
    \\"/2.0/repositories/{username}/{slug}/pullrequests/{pid}\\": {
        get: (context: 20RepositoriesUsernameSlugPullrequestsPidGetContext & TContext) => 20RepositoriesUsernameSlugPullrequestsPidGetResponse;
    };
    \\"/2.0/repositories/{username}/{slug}/pullrequests/{pid}/merge\\": {
        post: (context: 20RepositoriesUsernameSlugPullrequestsPidMergePostContext & TContext) => ResolverResponse;
    };
}

export interface user {
    username?: string;
    uuid?: string;
    [key: string]: any;
}

export type 20UsersUsernameGetResponse = (user | LaminarResponse<user>);

export interface 20UsersUsernameGetContext extends OapiContext, RouteContext {
    path: {
        username: string;
    };
}

export interface repository {
    slug?: string;
    owner?: user;
    [key: string]: any;
}

export type 20RepositoriesUsernameGetResponse = (repository[] | LaminarResponse<repository[]>);

export interface 20RepositoriesUsernameGetContext extends OapiContext, RouteContext {
    path: {
        username: string;
    };
}

export type 20RepositoriesUsernameSlugGetResponse = (repository | LaminarResponse<repository>);

export interface 20RepositoriesUsernameSlugGetContext extends OapiContext, RouteContext {
    path: {
        username: string;
        slug: string;
    };
}

export interface pullrequest {
    id?: number;
    title?: string;
    repository?: repository;
    author?: user;
    [key: string]: any;
}

export type 20RepositoriesUsernameSlugPullrequestsGetResponse = (pullrequest[] | LaminarResponse<pullrequest[]>);

export interface 20RepositoriesUsernameSlugPullrequestsGetContext extends OapiContext, RouteContext {
    path: {
        username: string;
        slug: string;
    };
    query: {
        state?: \\"open\\" | \\"merged\\" | \\"declined\\";
    };
}

export type 20RepositoriesUsernameSlugPullrequestsPidGetResponse = (pullrequest | LaminarResponse<pullrequest>);

export interface 20RepositoriesUsernameSlugPullrequestsPidGetContext extends OapiContext, RouteContext {
    path: {
        username: string;
        slug: string;
        pid: string;
    };
}

export interface 20RepositoriesUsernameSlugPullrequestsPidMergePostContext extends OapiContext, RouteContext {
    path: {
        username: string;
        slug: string;
        pid: string;
    };
}"
`;

exports[`Json Schema Ts Test petstore.yaml 1`] = `
"import { LaminarResponse, RouteContext } from \\"@ovotech/laminar\\";

import { OapiContext, OapiPaths, OapiSecurityResolver, OapiSecurityResolvers } from \\"@ovotech/laminar-oapi\\";

export interface Paths<TContext = {}> extends OapiPaths<TContext> {
    \\"/pets\\": {
        get: (context: PetsGetContext & TContext) => PetsGetResponse;
        post: (context: PetsPostContext & TContext) => PetsPostResponse;
    };
    \\"/pets/{petId}\\": {
        get: (context: PetsPetIdGetContext & TContext) => PetsPetIdGetResponse;
    };
}

export type Pets = Pet[];

export interface Pet {
    id: number;
    name: string;
    tag?: string;
    [key: string]: any;
}

export interface Error {
    code: number;
    message: string;
    [key: string]: any;
}

export type PetsGetResponse = (Pets | LaminarResponse<Pets>) | (Error | LaminarResponse<Error>);

export interface PetsGetContext extends OapiContext, RouteContext {
    query: {
        limit?: number;
    };
}

export type PetsPostResponse = (Error | LaminarResponse<Error>);

export interface PetsPostContext extends OapiContext, RouteContext {
}

export type PetsPetIdGetResponse = (Pets | LaminarResponse<Pets>) | (Error | LaminarResponse<Error>);

export interface PetsPetIdGetContext extends OapiContext, RouteContext {
    path: {
        petId: string;
    };
}"
`;

exports[`Json Schema Ts Test petstore-expanded.yaml 1`] = `
"import { LaminarResponse, RouteContext } from \\"@ovotech/laminar\\";

import { OapiContext, OapiPaths, OapiSecurityResolver, OapiSecurityResolvers } from \\"@ovotech/laminar-oapi\\";

export interface Paths<TContext = {}> extends OapiPaths<TContext> {
    \\"/pets\\": {
        /**
         * Returns all pets from the system that the user has access to
         * Nam sed condimentum est. Maecenas tempor sagittis sapien, nec rhoncus sem sagittis sit amet. Aenean at gravida augue, ac iaculis sem. Curabitur odio lorem, ornare eget elementum nec, cursus id lectus. Duis mi turpis, pulvinar ac eros ac, tincidunt varius justo. In hac habitasse platea dictumst. Integer at adipiscing ante, a sagittis ligula. Aenean pharetra tempor ante molestie imperdiet. Vivamus id aliquam diam. Cras quis velit non tortor eleifend sagittis. Praesent at enim pharetra urna volutpat venenatis eget eget mauris. In eleifend fermentum facilisis. Praesent enim enim, gravida ac sodales sed, placerat id erat. Suspendisse lacus dolor, consectetur non augue vel, vehicula interdum libero. Morbi euismod sagittis libero sed lacinia.
         *
         * Sed tempus felis lobortis leo pulvinar rutrum. Nam mattis velit nisl, eu condimentum ligula luctus nec. Phasellus semper velit eget aliquet faucibus. In a mattis elit. Phasellus vel urna viverra, condimentum lorem id, rhoncus nibh. Ut pellentesque posuere elementum. Sed a varius odio. Morbi rhoncus ligula libero, vel eleifend nunc tristique vitae. Fusce et sem dui. Aenean nec scelerisque tortor. Fusce malesuada accumsan magna vel tempus. Quisque mollis felis eu dolor tristique, sit amet auctor felis gravida. Sed libero lorem, molestie sed nisl in, accumsan tempor nisi. Fusce sollicitudin massa ut lacinia mattis. Sed vel eleifend lorem. Pellentesque vitae felis pretium, pulvinar elit eu, euismod sapien.
         *
         */
        get: (context: PetsGetContext & TContext) => PetsGetResponse;
        /**
         * Creates a new pet in the store.  Duplicates are allowed
         */
        post: (context: PetsPostContext & TContext) => PetsPostResponse;
    };
    \\"/pets/{id}\\": {
        /**
         * Returns a user based on a single ID, if the user does not have access to the pet
         */
        get: (context: PetsIdGetContext & TContext) => PetsIdGetResponse;
        /**
         * deletes a single pet based on the ID supplied
         */
        delete: (context: PetsIdDeleteContext & TContext) => PetsIdDeleteResponse;
    };
}

export type Pet = NewPet & {
    id: number;
    [key: string]: any;
};

export interface NewPet {
    name: string;
    tag?: string;
    [key: string]: any;
}

export interface Error {
    code: number;
    message: string;
    [key: string]: any;
}

export type PetsGetResponse = (Pet[] | LaminarResponse<Pet[]>) | (Error | LaminarResponse<Error>);

export interface PetsGetContext extends OapiContext, RouteContext {
    query: {
        tags?: string[];
        limit?: number;
    };
}

export type PetsPostResponse = (Pet | LaminarResponse<Pet>) | (Error | LaminarResponse<Error>);

export interface PetsPostContext extends OapiContext, RouteContext {
    body: NewPet;
}

export type PetsIdGetResponse = (Pet | LaminarResponse<Pet>) | (Error | LaminarResponse<Error>);

export interface PetsIdGetContext extends OapiContext, RouteContext {
    path: {
        id: number;
    };
}

export type PetsIdDeleteResponse = (Error | LaminarResponse<Error>);

export interface PetsIdDeleteContext extends OapiContext, RouteContext {
    path: {
        id: number;
    };
}"
`;

exports[`Json Schema Ts Test uspto.yaml 1`] = `
"import { LaminarResponse, RouteContext } from \\"@ovotech/laminar\\";

import { OapiContext, OapiPaths, OapiSecurityResolver, OapiSecurityResolvers } from \\"@ovotech/laminar-oapi\\";

export interface Paths<TContext = {}> extends OapiPaths<TContext> {
    \\"/\\": {
        get: (context: GetContext & TContext) => GetResponse;
    };
    \\"/{dataset}/{version}/fields\\": {
        /**
         * This GET API returns the list of all the searchable field names that are in the oa_citations. Please see the 'fields' attribute which returns an array of field names. Each field or a combination of fields can be searched using the syntax options shown below.
         */
        get: (context: DatasetVersionFieldsGetContext & TContext) => DatasetVersionFieldsGetResponse;
    };
    \\"/{dataset}/{version}/records\\": {
        /**
         * This API is based on Solr/Lucense Search. The data is indexed using SOLR. This GET API returns the list of all the searchable field names that are in the Solr Index. Please see the 'fields' attribute which returns an array of field names. Each field or a combination of fields can be searched using the Solr/Lucene Syntax. Please refer https://lucene.apache.org/core/3_6_2/queryparsersyntax.html#Overview for the query syntax. List of field names that are searchable can be determined using above GET api.
         */
        post: (context: DatasetVersionRecordsPostContext & TContext) => DatasetVersionRecordsPostResponse;
    };
}

export interface dataSetList {
    total?: number;
    apis?: {
        /**
         * To be used as a dataset parameter value
         */
        apiKey?: string;
        /**
         * To be used as a version parameter value
         */
        apiVersionNumber?: string;
        /**
         * The URL describing the dataset's fields
         */
        apiUrl?: string;
        /**
         * A URL to the API console for each API
         */
        apiDocumentationUrl?: string;
        [key: string]: any;
    }[];
    [key: string]: any;
}

export type GetResponse = (dataSetList | LaminarResponse<dataSetList>);

export interface GetContext extends OapiContext, RouteContext {
}

export type DatasetVersionFieldsGetResponse = (string | LaminarResponse<string>) | LaminarResponse<string>;

export interface DatasetVersionFieldsGetContext extends OapiContext, RouteContext {
    path: {
        dataset: string;
        version: string;
    };
}

export type DatasetVersionRecordsPostResponse = (any[] | LaminarResponse<any[]>);

export interface DatasetVersionRecordsPostContext extends OapiContext, RouteContext {
    path: {
        version: string;
        dataset: string;
    };
    body?: any;
}"
`;
